src/tools/replay.cpp

```diff
@@ -79,30 +82,25 @@ void record_action(std::string action_type, std::map<std::string,std::string> in
 }
 
 bool has_next_action() {
-	Element* root = log_document.FirstChildElement();
-	return root->FirstChildElement(false) != NULL;
+	return next_action != NULL;
 }
 
 std::string next_action_type() {
-	Element* root = log_document.FirstChildElement();
-	Element* next_action = root->FirstChildElement();
 	return next_action->Value();
 }
 
 Element* pop_next_action(std::string expected_action_type) {
-	Element* root = log_document.FirstChildElement();
-	Element* next_action = root->FirstChildElement();
-	
 	if (expected_action_type != "" && next_action->Value() != expected_action_type) {
 		std::ostringstream stream;
 		stream << "Replay error! Expected '" << expected_action_type << "' action next";
 		throw stream.str();
 	}
 
-	Element* clone = next_action->Clone()->ToElement();
-	root->RemoveChild(next_action);
+	Element* to_return = next_action;
+	
+	next_action = next_action->NextSiblingElement(false);
 	
-	return clone;
+	return to_return;
```

Comment:
Is it intended that pop_next_action can return null? If it's never null, it should be returned as a reference, but if the expectation is that it could return null, then this is fine as is.

Reply:
It's intended that you call has_next_action() before popping, so returning Element& makes sense